### ISSUE TITLE
How to deal with multiple data sets?

### ISSUE THREAD

**Issue by ziyinyuan** [dynamicslab/pysindy#236]
I have multiple data sets in time, and each data set is a function of time and parameters f_i (\alpha_i, t). I would like to discover the governing equations such that I can predict f_j (\alpha_j, t), where f_j is not part of the training data.

To put my problem in a simpler case, I try:
a=2
b=3
t=np.arange(0,10,0.01)
x1=cos(a*t)
x2=cos(b*t)
x=np.concatenate((x1,x2),axis=0).reshape((-1,1)) to combine x1 and x2 in 1 column.

Am I starting on the right track? I can get governing equation if I start off this way, but it takes forever to simulate the result.

**Comment by akaptano** [comment]
Looks fine so far. Please post a plot of the fit of x_dot (using model.predict) and x (using model.simulate), as well as the model.print() output. If it is taking forever to simulate, there is probably numerical instability in the model, but hard to say without more information from you.

**Comment by ziyinyuan** [comment]
Here's the code I used for testing:

---------------------------------------------------------------------------------
dt=0.01
t=np.arange(0,5,dt)
a=3
b=5

x1=np.cos(a*t)
x2=np.cos(b*t)
x=np.concatenate((x1,x2),axis=0).reshape((-1,1))

time_oneset=len(t_2d)
t_2d=np.concatenate((t,t),axis=0).reshape((-1,1))
a_2d=np.tile(a,time_oneset).reshape((-1,1))
b_2d=np.tile(b,time_oneset).reshape((-1,1))
u1=np.concatenate((a_2d,b_2d,t_2d),axis=1)

f1 = lambda alpha, t: np.cos(a * t)
f1_name = lambda alpha, t: f"(cos({a}*{t}))"
f2 = lambda alpha, t: np.sin(a * t)
f2_name = lambda alpha, t: f"(sin({a}*{t}))"
f3 = lambda alpha, t: np.cos(b * t)
f3_name = lambda alpha, t: f"(cos({b}*{t}))"
f4 = lambda alpha, t: np.sin(b * t)
f4_name = lambda alpha, t: f"(sin({b}*{t}))"

# Only Custom Lib
lib=ps.CustomLibrary(
    library_functions=[f1,f2,f3,f4],
    function_names=[f1_name,f2_name,f3_name,f4_name]
)
generalized_library=ps.GeneralizedLibrary(
    libraries=[lib],
    inputs_per_library=np.array([[1,1,3,3]])
)

t1=np.arange(0,10,dt)
model=ps.SINDy(
    differentiation_method=ps.FiniteDifference(order=2),
    optimizer=ps.STLSQ(threshold=0.002),
    feature_names=["x","a","b","t"],
    feature_library=generalized_library
               )
model.fit(x,t=t1,u=u1)
model.print()


t_test=np.arange(0,5,0.01)
x0=x[0]
a_test=3
b_test=5

t_test_2d=t_test.reshape((-1,1))
a_test_2d=a_test*np.ones_like(t_test_2d)
b_test_2d=b_test*np.ones_like(t_test_2d)
x_test=np.concatenate((a_test_2d,b_test_2d,t_test_2d),axis=1)

x0_sim=model.simulate(x0,t1,u1)
#**there's an UserWarning said "Last time point dropped in simulation because interpolation of control input was used. To avoid this, pass in a callable for u.**

t_test_plot=np.arange(0,10,0.01)
t_test_plot1=np.arange(0,5,0.01)
t_test_plot2=np.arange(5,10,0.01)
plt.figure()
plt.plot(t_test_plot1,x1,label="cos(a*t)",linewidth=4)
plt.plot(t_test_plot2,x2,label="cos(b*t)",linewidth=4)
plt.plot(t_test_plot[0:999],x0_sim,"k--",label="SINDy model",linewidth=3)
plt.legend()
plt.xlabel("t")
plt.ylabel("x")
![image](https://user-images.githubusercontent.com/103147478/182004366-9f7fd985-f345-47d2-8401-1c719767581c.png)

By using only the custom library, I can simulate but got an incorrect result. Please correct me if there are any mistakes I've made.

**Comment by ziyinyuan** [comment]
If I add in Polynomial Library for testing:

----------------------------------------------------------------------------------------------

inputs_temp=np.tile([0,1,2,3],2)
inputs_per_lib=np.reshape(inputs_temp,(2,4))

lib=ps.CustomLibrary(
    library_functions=[f1,f2,f3,f4],
    function_names=[f1_name,f2_name,f3_name,f4_name]
)
poly_lib=ps.PolynomialLibrary(degree=2)
generalized_library=ps.GeneralizedLibrary(
    libraries=[poly_lib,lib],
    inputs_per_library=inputs_per_lib
)

t1=np.arange(0,10,dt)
model=ps.SINDy(
    differentiation_method=ps.FiniteDifference(order=2),
    optimizer=ps.STLSQ(threshold=0.002),
    feature_names=["x","a","b","t"],
    feature_library=generalized_library
               )
model.fit(x,t=t1,u=u1)
model.print()

t_test=np.arange(0,5,0.01)
x0=x[0]
a_test=3
b_test=5

t_test_2d=t_test.reshape((-1,1))
a_test_2d=a_test*np.ones_like(t_test_2d)
b_test_2d=b_test*np.ones_like(t_test_2d)
x_test=np.concatenate((a_test_2d,b_test_2d,t_test_2d),axis=1)

x0_sim=model.simulate(x0,t1,u1)

------------------------------------------------------------------------

My program was then stuck in running the simulation.

**Comment by Jacob-Stevens-Haas** [comment]
@ziyinyuan , I've edited your comment to show how to format code in github.

It soulds like you want an equation
$$x(a, t) = \cos (a t)$$

This is a function of both `a` and `t`, and it's derivative involves both as well.  Try adding this to your `CustomLibrary`:
```python
f1 = lambda a, t: sin(a, t)
```
You will still need `a` and `t` in your control input `u`.

**Comment by ziyinyuan** [comment]
Thank you so much for correcting the function. 

---------------------------------------------------------------------

f1 = lambda a, t: np.cos(a * t)
f1_name = lambda a, t: f"(cos({a}*{t}))"
f2 = lambda a, t: np.sin(a * t)
f2_name = lambda a, t: f"(sin({a}*{t}))"

f3 = lambda b, t: np.cos(b * t)
f3_name = lambda b, t: f"(cos({b}*{t}))"
f4 = lambda b, t: np.sin(b * t)
f4_name = lambda b, t: f"(sin({b}*{t}))"

inputs_temp=np.tile([0,1,2,3],3)
inputs_per_lib=np.reshape(inputs_temp,(3,4))

inputs_per_lib[0,0]=1
inputs_per_lib[0,1]=1
inputs_per_lib[0,2]=2
inputs_per_lib[0,3]=3

inputs_per_lib[1,0]=1
inputs_per_lib[1,1]=1
inputs_per_lib[1,2]=3
inputs_per_lib[1,3]=3

inputs_per_lib[2,0]=2
inputs_per_lib[2,1]=2
inputs_per_lib[2,2]=3
inputs_per_lib[2,3]=3

lib1=ps.CustomLibrary(
    library_functions=[f1,f2],
    function_names=[f1_name,f2_name]
)
lib2=ps.CustomLibrary(
    library_functions=[f3,f4],
    function_names=[f3_name,f4_name]
)
poly_lib=ps.PolynomialLibrary(degree=2)
generalized_library=ps.GeneralizedLibrary(
    libraries=[poly_lib,lib1,lib2],
    inputs_per_library=inputs_per_lib
)
t1=np.arange(0,len(t_2d)*dt,dt)
model=ps.SINDy(
    differentiation_method=ps.FiniteDifference(order=2),
    optimizer=ps.STLSQ(threshold=0.002),
    feature_names=["x","a","b","t"],
    feature_library=generalized_library
               )
model.fit(x,t=t1,u=u1)
model.print()

-----------------------------------------------------------------------

The equation came out to be:
(x)' = -1.000 (sin(a*t)) + -1.999 (sin(b*t))

The amplitude doesn't look correct. It should be 
(x)' = -2.000 (sin(a*t)) + -4 (sin(b*t)).

Is it because of the customized library or are there any mistakes I made?

**Comment by Jacob-Stevens-Haas** [comment]
Please format your code with 
\`\`\`python

type your code here

\`\`\`

For more information on formatting in github, see the [markdown guide](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code).  For more detail on code blocks, see [this link](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks).

I did not get the same results as you.  It might be because I generated different data.  Please post a _complete_ example that runs and produces that result.

**Comment by ziyinyuan** [comment]
Here my complete code:

```python

dt=0.01
t=np.arange(0,3.1415*5,dt)

a=2
b=4

x1=np.cos(a*t)
x2=np.cos(b*t)
x=np.concatenate((x1,x2),axis=0).reshape((-1,1))
length_twoset=len(x)

t_2d = np.concatenate((t, t), axis=0).reshape((-1, 1)) 
a_2d=np.tile(a,length_twoset).reshape((-1,1))
b_2d=np.tile(b,length_twoset).reshape((-1,1))
u1=np.concatenate((a_2d,b_2d,t_2d),axis=1)

f1 = lambda a, t: np.cos(a * t)
f1_name = lambda a, t: f"(cos({a}*{t}))"
f2 = lambda a, t: np.sin(a * t)
f2_name = lambda a, t: f"(sin({a}*{t}))"
f3 = lambda b, t: np.cos(b * t)
f3_name = lambda b, t: f"(cos({b}*{t}))"
f4 = lambda b, t: np.sin(b * t)
f4_name = lambda b, t: f"(sin({b}*{t}))"

inputs_temp=np.tile([0,1,2,3],3)
inputs_per_lib=np.reshape(inputs_temp,(3,4))

inputs_per_lib[0,0]=1
inputs_per_lib[0,1]=1
inputs_per_lib[0,2]=2
inputs_per_lib[0,3]=3

inputs_per_lib[1,0]=1
inputs_per_lib[1,1]=1
inputs_per_lib[1,2]=3
inputs_per_lib[1,3]=3

inputs_per_lib[2,0]=2
inputs_per_lib[2,1]=2
inputs_per_lib[2,2]=3
inputs_per_lib[2,3]=3

lib1=ps.CustomLibrary(
    library_functions=[f1,f2],
    function_names=[f1_name,f2_name]
)
lib2=ps.CustomLibrary(
    library_functions=[f3,f4],
    function_names=[f3_name,f4_name]
)
poly_lib=ps.PolynomialLibrary(degree=2)
generalized_library=ps.GeneralizedLibrary(
    libraries=[poly_lib,lib1,lib2],
    inputs_per_library=inputs_per_lib
)

t1=np.arange(0,len(t_2d)*dt,dt)
model=ps.SINDy(
    differentiation_method=ps.FiniteDifference(order=2),
    optimizer=ps.STLSQ(threshold=0.002),
    feature_names=["x","a","b","t"],
    feature_library=generalized_library
               )
model.fit(x,t=t1,u=u1)
model.print()

```

**Comment by Jacob-Stevens-Haas** [comment]
Thanks!  This line is the culprit:
```
x1=np.cos(a*t)
x2=np.cos(b*t)
x=np.concatenate((x1,x2),axis=0).reshape((-1,1))
```
Here you are stacking your data on top of each other in time.  That is, `x` behaves for the first half like `np.cos(a * t)` and the second half like `np.cos(b * t)`.  So SINDy is finding the average of those two models.

**Comment by ziyinyuan** [comment]
What if I only have the data for x1 and x2 and the constant for a and b instead of the equation?

**Comment by Jacob-Stevens-Haas** [comment]
I'm not sure what you mean.  The point of SINDy is to find the equation.  In general, you either need to know something about the dynamics or start by trying out a polynomial library.

**Comment by ziyinyuan** [comment]
I was trying to say, what if I start with the data for x1 and x2 and the constant for a and b instead of equations (cos(at) and cos(bt))? Originally, I was thinking to concatenate the x1 and x2 into 1 column, but you mentioned that the SINDy will find the average of the 2 models. So do you have any suggestions on how would I start to construct the x?

**Comment by Jacob-Stevens-Haas** [comment]
Are you saying you don't know what `x` is?  All of sindy requires you to know what `x` is.  Then you choose some possible functions for $\dot x$ and SINDy finds which of the functions fit the data.

**Comment by ziyinyuan** [comment]
I was trying to say, for:
```python
x1=np.cos(a*t)
x2=np.cos(b*t)
```
if 
```python
t=np.arange(0,3.1415*5,0.1)
a=2
b=3
```
I will have the x1 and x2 data.

If I want to train both equations together with the parameters of 'a' and 'b', I can add the two functions together to get 'x':
```python
x=np.cos(a*t)+np.cos(b*t)
```

So my question now would be if I know x1, x2, (which are the 2 columns of data for cos(at) and cos(bt)), a, b, and t, but let's assume that I don't know the function for x1 and x2. How should I get the 'x' to start training the data?

**Comment by akaptano** [comment]
I'm going to close this as being fairly off-topic by now from the original question. ziyinyuan please take a look at the Example 1 Jupyter notebook for how to properly use the code.

**Comment by Jacob-Stevens-Haas** [comment]
@ziyinyuan try concatenating x1 and x2 along axis=1.
